flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecDeduplicate.java

```diff
@@ -339,25 +362,39 @@ OneInputStreamOperator<RowData, RowData> createDeduplicateOperator() {
                 }
             } else {
                 if (isAsyncStateEnabled()) {
-                    AsyncStateRowTimeDeduplicateFunction processFunction =
-                            new AsyncStateRowTimeDeduplicateFunction(
-                                    rowTypeInfo,
-                                    stateRetentionTime,
-                                    rowtimeIndex,
-                                    generateUpdateBefore,
-                                    generateInsert(),
-                                    keepLastRow);
-                    return new AsyncKeyedProcessOperator<>(processFunction);
+                    if (!keepLastRow && outputInsertOnly) {
+                        checkState(canBeInsertOnly(config, keepLastRow));
```

Comment:
nit: can we need to double check the field `outputInsertOnly`?
The same goes for the following.

Reply:
Generating plans and actually using those plans can happen quite far apart. Yes it doesn't make much sense to `checkState` this if we assume both generation and creating job graph happens with the same Flink version, but that's often not the case. With those `checkState` I wanted to add a bit of extra safety if in the future someones changes the code. Similarly how for example the dedupe operator itself makes assertions that the input is insert only.

Also they don't hurt.

But if you have stronger feelings, about dropping them I would be fine as well.

Reply:
Emmm... I meant to add one more assertion to do what you say, as shown below to ensure consistency between visiting FlinkChangelogModeInferenceProgram during optimizing and executing this plan. WDYT?
```
checkState(!RankUtil.outputInsertOnlyInDeduplicate(config, keepLastRow));
checkState(!outputInsertOnly)
```

Reply:
Ahh, sorry I misunderstood you.

You mean why don't I add the reverse check state in the else branch below? The problem with that is that old compiled plans would fail that check state. I mean plans that were created before this PR, but are executed with this PR. `RankUtil.outputInsertOnlyInDeduplicate(config, keepLastRow)` could return `true` for them, while `outputInsertOnly` is false (absent value/null interpreted as false)

Reply:
Maybe I didn't make it clear, I just want to add one more check in the `if` branch like:
```
if (isMiniBatchEnabled()) {
    checkState(!RankUtil.outputInsertOnlyInDeduplicate(config, keepLastRow));
    checkState(!outputInsertOnly);
    ...
} else {
    ...
}
```

Reply:
Ahh, ok. Done